mixer/pkg/checkcache/cache.go

```diff
@@ -0,0 +1,202 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package checkcache provides a scalable cache to hold results of Mixer.Check operations.
+//
+// Entries are added into the cache by supplying an attribute bag along with a ReferencedAttributes struct
+// which determines the set of attributes in the bag should be used as a cache lookup key. Entries are looked up
+// from the cache using an attribute bag.
+package checkcache
+
+// TODO: This code should optimize the storage of CacheValue. It's likely that a great many entries in the cache will
+// have identical StatusMessage and ReferencedAttributes values and sharing these immutable objects across
+// entries would save a substantial amount of storage relative to the whole cache size.
+//
+// TODO: should there be a ReferencedAttribute struct associated with each KeyShape so that it doesn't need to be stored
+//       with each corresponding cached value?
+
+import (
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	mixerpb "istio.io/api/mixer/v1"
+	"istio.io/istio/mixer/pkg/attribute"
+	"istio.io/istio/pkg/cache"
+)
+
+// CheckCache holds cache results of calls to Mixer.Check
+type CheckCache struct {
```

Comment:
I don't have a ton of suggestions, but there is a lot of stutter here: `checkcache.CheckCache`.  Having the package just be `check` (and then `check.Cache`) is attractive in some ways, but misleading in others.

Reply:
I was hoping you'd have some great ideas here :-)

Reply:
I don't have any good suggestions sadly. We could put this in a `perf` package, but that's probably too obtuse. `check` implies that other bits are here than just cache stuff.

Maybe we just change `CacheValue` to `Value` and `CheckCache` to just `Cache` (or something awful like `Instance`)?